---
title: "Chelicerate Hox genes"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(cowplot)
#library(ggbeeswarm)
library(dplyr)
library(kableExtra)
#library(tidyr)
library(ggtree)
#library(phytools)
#library(phangorn)
#library(reshape2)
library(ggExtra)
#library(ggrepel)
#library(vroom)
#library(ggdist)
#library(stringr)
#library(ggsignif)
library(here)
source("../lib/design.r")
source("../lib/get_tree_info.r")
#source("C:/Users/grt814/bin/core/corelib/design.r")
#source("C:/Users/grt814/bin/core/get_tree_info.r")

#htmltools::includeHTML("../html-chunks/rmd_nav.html")
```

[< Back to project](index.html)

Gene trees for genes from the [Hox gene complex from the Drosophila annotation](https://flybase.org/reports/FBgg0000363.html). Orthologs identified with FastOrtho and alignments filtered according to [this](https://gwct.github.io/spider-wgd/alignments.html).

```{r file-setup, out.width="75%", fig.align = "center", warning=FALSE}

hox_data_file = here("data", "dmela-hox-genes.csv")
hox_data = read.csv(hox_data_file, header=T, comment.char="#")


```


```{r hox-figs, out.width="66%", fig.align = "center", warning=FALSE}

for(i in 1:nrow(hox_data)){
  row = hox_data[i,]
  
  title = paste(row$Gene, row$Gene.id, row$Protein.chosen, row$Orthogroup, sep=" / ")
  
  cur_tree = read.tree(text=row$CDS.tree)
  
  cur_gt_fig = ggtree(cur_tree, size=0.8, ladderize=T) +#, aes(color=full_data$Family)) +
  xlim(0, 4) +
  geom_tiplab(size=4) +
  #scale_color_manual(name='Include?', values=c("N"=corecol(pal="wilke", numcol=1,offset=1), "Y"=corecol(numcol=1, offset=1))) +
  #geom_text(aes(label=node), hjust=-.3, vjust=-.3, color="#ff6db6") +
  
  # geom_cladelabel(node=22, label="Araneae", align=T, color=corecol(numcol=1, offset=2), offset=famoffset) +
  # geom_cladelabel(node=7, label="Scorpiones", align=T, color=corecol(numcol=1), offset=famoffset) +
  # geom_cladelabel(node=27, label="Xiphosura", align=T, color=corecol(numcol=1, offset=4), offset=famoffset) +
  # geom_cladelabel(node=28, label="Acariformes", align=T, color=corecol(numcol=1, offset=7), offset=famoffset) +
  # geom_cladelabel(node=29, label="Parasitiformes", align=T, color=corecol(numcol=1, offset=1, pal="trek"), offset=famoffset) +
  # geom_cladelabel(node=15, label="Diptera", align=T, color="#333333", offset=famoffset) +
  # geom_cladelabel(node=16, label="Lepidoptera", align=T, color="#333333", offset=famoffset) +
  ggtitle(title)
  
  print(cur_gt_fig)
  
  if(!is.na(row$Notes)){
    print(row$Notes)
  }
  
  print("-------------------------------")
  #stop()
}

```

[< Back to project](index.html)