---
title: "Chelicerate species trees"
author: "[gwct](https://gwct.github.io/)"
date: "`r format(Sys.time(), '%m/%d/%Y %H:%M:%S %Z')`"
output:
  #html_document:
  rmdformats::robobook:
    highlight: kate
    includes:
      in_header: '../html-chunks/rmd_header.html'
    df_print: paged
    code_folding: hide
    number_sections: true
    toc_depth: 3
---

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(ggplot2)
library(cowplot)
#library(ggbeeswarm)
library(dplyr)
library(kableExtra)
#library(tidyr)
library(ggtree)
#library(phytools)
library(phangorn)
#library(reshape2)
library(ggExtra)
#library(ggrepel)
#library(vroom)
#library(ggdist)
#library(stringr)
#library(ggsignif)
library(here)
source("../lib/design.r")
source("../lib/get_tree_info.r")
#source("C:/Users/grt814/bin/core/corelib/design.r")
#source("C:/Users/grt814/bin/core/get_tree_info.r")

#ggtree::rotate <- ggtree::rotate()

#htmltools::includeHTML("../html-chunks/rmd_nav.html")

prune_tips <- function(tree1, tree2) {
# Prunes tips from tree2 that are not in tree 1
# Returns pruned tree
  
  tips_to_prune = tree2$tip.label[!tree2$tip.label %in% tree1$tip.label]
  
  if(length(tips_to_prune) > 0){
    tree2 = drop.tip(tree2, tips_to_prune)
  }
  return(tree2)
}

```

[< Back to project](index.html)

```{r file-setup, out.width="75%", fig.align = "center", warning=FALSE}
save_tree_fig = F
# Meta options. Comment tree_type when running form generator

genome_file = here("data", "chelicerate_genomes-gwct.csv")

genome_data = read.csv(genome_file, header=T)
genome_data = subset(genome_data, Dataset!="Exclude")

spec_abbrs = select(genome_data, S.name, Abbr, Superorder, Order)
names(spec_abbrs)[2] = "label"

# Genome info
####################

summary_file_16 = here("data", "16spec", "gene-trees.csv")
summary_file_19 = here("data", "19spec", "gene-trees.csv")

# Gene tree summaries
####################

pro_file_16 = here("data", "16spec", "trees", "chelicerate-16spec-astral-pro.rooted.tre")
pro_file_19 = here("data", "19spec", "trees", "chelicerate-19spec-astral-pro.rooted.tre")

multi_file_16 = here("data", "16spec", "trees", "chelicerate-16spec-astral-multi.rooted.tre")
multi_file_19 = here("data", "19spec", "trees", "chelicerate-19spec-astral-multi.rooted.tre")

sc_file_16 = here("data", "16spec", "trees", "chelicerate-16spec-astral-single-copy.rooted.tre")
sc_file_19 = here("data", "19spec", "trees", "chelicerate-19spec-astral-single-copy.rooted.tre")

bal_file_16 = here("data", "16spec", "trees", "ballesteros.tre")
bal_file_19 = here("data", "19spec", "trees", "ballesteros.tre")

trad_file_16 = here("data", "16spec", "trees", "traditional.tre")
trad_file_19 = here("data", "19spec", "trees", "traditional.tre")

# Species trees
####################

```

# Species and taxonomy lookup

```{r spec-summary, out.width="75%", fig.align = "center", warning=FALSE}

spec_abbrs %>% kable() %>% kable_styling(bootstrap_options=c("striped", "condended", "responsive"), full_width=F)
```


# Gene tree summary

Note that status refers only to the rooting process. All 9629 gene trees were used for ASTRAL-pro and ASTRAL-multi.

```{r gt-summary, out.width="75%", fig.align = "center", warning=FALSE}
gt_data = read.csv(summary_file_16, header=T)
#gt_data_16$dataset = "16 species"
gt_data_19 = read.csv(summary_file_19, header=T)
#gt_data_19$dataset = "19 species"
names(gt_data)[3] = "Count.16.spec"
gt_data$Count.19.spec = gt_data_19$Count

gt_data %>% kable() %>% kable_styling(bootstrap_options=c("striped", "condended", "responsive"), full_width=F)
```

# ASTRAL-pro species tree

Inferred from gene trees using the C++ implementation of ASTRAL-pro (https://github.com/chaoszhang/ASTER).

All branches have full ASTRAL support (1.0).

Note: Tip branch lengths arbitrarily set to average of internal branch lengths...

## 16 species tree vs. 19 species tree

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-pro.rooted.tre) (RIGHT)**

```{r astral-pro-tree-cds, out.width="35%", fig.align = "center", warning=FALSE}

orig_pro_tree_16 = read.tree(file=pro_file_16)
tree_to_df_list = treeToDF(orig_pro_tree_16, add_avg_bl=TRUE)
pro_info_16 = tree_to_df_list[["info"]]
pro_tree_16 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

orig_pro_tree_19 = read.tree(file=pro_file_19)
tree_to_df_list = treeToDF(orig_pro_tree_19, add_avg_bl=TRUE)
pro_info_19 = tree_to_df_list[["info"]]
pro_tree_19 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(pro_tree_16, pro_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

pro_tree_19_pruned = prune_tips(pro_tree_16, pro_tree_19)
#RF.dist(pro_tree_16, pro_tree_19_pruned)

comparePhylo(pro_tree_16, pro_tree_19_pruned, commons=F, plot=T, location=NA)

```

**Notably, Acariformes (SSCAB, TURTI) switch places with the horseshoe crabs in the 19 species tree, putting them sister to spiders and scorpions**

## 16 species tree vs. Ballesteros species tree

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (Ballesteros topology; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/ballesteros.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-pro-tree-cds-16-bal, out.width="35%", fig.align = "center", warning=FALSE}

orig_bal_tree_16 = read.tree(file=bal_file_16)
tree_to_df_list = treeToDF(orig_bal_tree_16, add_avg_bl=TRUE)
bal_info_16 = tree_to_df_list[["info"]]
bal_tree_16 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(pro_tree_16, bal_tree_16)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_16, bal_tree_16)
comparePhylo(pro_tree_16, bal_tree_16, commons=F, plot=T, location=NA)

```


## 19 species tree vs. Ballesteros species tree

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (Ballesteros topology; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/ballesteros.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-pro-tree-cds-19-bal, out.width="35%", fig.align = "center", warning=FALSE}

orig_bal_tree_19 = read.tree(file=bal_file_19)
tree_to_df_list = treeToDF(orig_bal_tree_19, add_avg_bl=TRUE)
bal_info_19 = tree_to_df_list[["info"]]
bal_tree_19 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(pro_tree_19, bal_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_19, bal_tree_19)
comparePhylo(pro_tree_19, bal_tree_19, commons=F, plot=T, location=NA)

```

## 16 species tree vs. Traditional species tree (horseshoe crabs not with arachnids)

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (HC non-arachnid; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/traditional.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-pro-tree-cds-16-trad, out.width="35%", fig.align = "center", warning=FALSE}

orig_trad_tree_16 = read.tree(file=trad_file_16)
tree_to_df_list = treeToDF(orig_trad_tree_16, add_avg_bl=TRUE)
trad_info_16 = tree_to_df_list[["info"]]
trad_tree_16 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(pro_tree_16, trad_tree_16)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=2), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_16, trad_tree_16)
comparePhylo(pro_tree_16, trad_tree_16, commons=F, plot=T, location=NA)

```

## 19 species tree vs. Traditional species tree (horseshoe crabs not with arachnids)

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (HC non-arachnid; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/traditional.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-pro-tree-cds-19-trad, out.width="35%", fig.align = "center", warning=FALSE}

orig_trad_tree_19 = read.tree(file=trad_file_19)
tree_to_df_list = treeToDF(orig_trad_tree_19, add_avg_bl=TRUE)
trad_info_19 = tree_to_df_list[["info"]]
trad_tree_19 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(pro_tree_19, trad_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=2), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_19, trad_tree_19)
comparePhylo(pro_tree_19, trad_tree_19, commons=F, plot=T, location=NA)

```

# ASTRAL-multi species tree

Inferred from gene trees using ASTRAL, and the gene-to-species mapping method outlined here: https://github.com/smirarab/ASTRAL/blob/master/astral-tutorial.md#memory-issues-due-to-taxon-names.

All branches have full ASTRAL support (1.0).

Note: Tip branch lengths inferred by ASTRAL since this is treating the paralogs like multiple individuals in a species

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-multi.rooted.tre)**

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-multi.rooted.tre)**

```{r astral-multi-tree-cds, out.width="35%", fig.align = "center", warning=FALSE}

multi_tree_16 = read.tree(file=multi_file_16)
multi_tree_19 = read.tree(file=multi_file_19)
# Read the trees

#comp = cophylo(orig_multi_tree_16, orig_multi_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(orig_multi_tree_16, prune_tips(orig_multi_tree_16, orig_multi_tree_19))
multi_tree_19_pruned = prune_tips(multi_tree_16, multi_tree_19)
comparePhylo(multi_tree_16, multi_tree_19_pruned, commons=F, plot=T, location=NA)

```

## 16 species tree vs. Ballesteros species tree

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-multi.rooted.tre) (LEFT)**

**[Newick tree file (Ballesteros topology; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/ballesteros.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-multi-tree-cds-16-bal, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(orig_multi_tree_16, bal_tree_16)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(orig_multi_tree_16, bal_tree_16)
comparePhylo(multi_tree_16, bal_tree_16, commons=F, plot=T, location=NA)

```

## 19 species tree vs. Ballesteros species tree

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-multi.rooted.tre) (LEFT)**

**[Newick tree file (Ballesteros topology; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/ballesteros.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-multi-tree-cds-19-bal, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(orig_multi_tree_19, bal_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(orig_multi_tree_19, bal_tree_19)
comparePhylo(multi_tree_19, bal_tree_19, commons=F, plot=T, location=NA)

```

## 16 species tree vs. Traditional species tree (horseshoe crabs not with arachnids)

**[Newick tree file (16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-multi.rooted.tre) (LEFT)**

**[Newick tree file (HC non-arachnid; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/traditional.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-multi-tree-cds-16-trad, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(orig_multi_tree_16, trad_tree_16)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=2), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(orig_multi_tree_16, trad_tree_16)
comparePhylo(multi_tree_16, trad_tree_16, commons=F, plot=T, location=NA)

```

## 19 species tree vs. Traditional species tree (horseshoe crabs not with arachnids)

**[Newick tree file (19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-multi.rooted.tre) (LEFT)**

**[Newick tree file (HC non-arachnid; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/traditional.tre) (RIGHT; arbitrary branch lengths)**

```{r astral-multi-tree-cds-19-trad, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(orig_multi_tree_19, trad_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=2), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(orig_multi_tree_19, trad_tree_19)
comparePhylo(multi_tree_19, trad_tree_19, commons=F, plot=T, location=NA)

```

# Pro vs. Multi trees

## 16 species

**[Newick tree file (PRO; 16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (MULTI; 16 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/16spec/trees/chelicerate-16spec-astral-multi.rooted.tre) (RIGHT)**

```{r pro-multi-16, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(pro_tree_16, orig_multi_tree_16)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_16, orig_multi_tree_16)
comparePhylo(pro_tree_16, multi_tree_16, commons=F, plot=T, location=NA)

```

## 19 species

**[Newick tree file (PRO; 19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-pro.rooted.tre) (LEFT)**

**[Newick tree file (MULTI; 19 species; CDS; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-multi.rooted.tre) (RIGHT)**

```{r pro-multi-19, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(pro_tree_19, orig_multi_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(pro_tree_19, orig_multi_tree_19)
comparePhylo(pro_tree_19, multi_tree_19, commons=F, plot=T, location=NA)

```

# ASTRAL species tree from single-copy orthologs

Inferred from single-copy orthologs using ASTRAL.

All branches have full ASTRAL support (1.0).

Note: Tip branch lengths arbitrarily set to average of internal branch lengths...

**[Newick tree file (19 species; CDS; 21 genes; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-single-copy.rooted.tre)**

**[Newick tree file (19 species; CDS; 8 genes; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/chelicerate-19spec-astral-single-copy.rooted.tre)**

```{r astral-sc-tree-cds, out.width="35%", fig.align = "center", warning=FALSE}

orig_sc_tree_16 = read.tree(file=sc_file_16)
tree_to_df_list = treeToDF(orig_sc_tree_16, add_avg_bl=TRUE)
sc_info_16 = tree_to_df_list[["info"]]
sc_tree_16 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

orig_sc_tree_19 = read.tree(file=sc_file_19)
tree_to_df_list = treeToDF(orig_sc_tree_19, add_avg_bl=TRUE)
sc_info_19 = tree_to_df_list[["info"]]
sc_tree_19 = tree_to_df_list[["labeled.tree"]]
# # Parse the tree info, and replace the tip branch lengths
# # with the avg. internal branch lengths as "labeled.tree"

#comp = cophylo(sc_tree_16, sc_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=1), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(sc_tree_16, prune_tips(sc_tree_16, sc_tree_19))
sc_tree_19_pruned = prune_tips(sc_tree_16, sc_tree_19)
comparePhylo(sc_tree_16, sc_tree_19_pruned, commons=F, plot=T, location=NA)

```

# Notes relating to Ballesteros et al. 2022

https://academic.oup.com/mbe/article/39/2/msac021/6522129

This paper samples a wide range of Chelicerate genomes and transcriptomes (516 samples) to try and resolve the relationships withing Chelicerata (relevant groups: Araneae, Scorpiones, Acariformes, Parasitiformes) and between Chelicerata and horseshoe crabs (Xiphosura). 

They find:

1. Xiphosura consistently groups with Araneae and Scorpiones
2. Acariformes (TURTI, SSCAB) and Parasitiformes (MOCCI, VDEST, HLONG, ISCAP) are not monophyletic with each other. Their full tree (Fig. 1, Matrix 3) puts Acariformes outside of all other Parasitiformes, Xiphosura, Scorpiones, and Araneae.
3. Using their most stringent dataset (Matrix 4), Acariformes is sister to Xiphosura, but the lengths of the branches separating these groups from each other and from Araneae and Scorpiones are very short (Fig. 3).

**[Newick tree file (HC non-arachnid; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/traditional.tre) (LEFT; arbitrary branch lengths)
**[Newick tree file (Ballesteros topology; rooted)](https://github.com/gwct/spider-wgd/blob/main/data/19spec/trees/ballesteros.tre) (RIGHT; arbitrary branch lengths)

Our results also consistently show Xiphosura sister to Araneae and Scopriones, within Arachnida. We also find inconsistent placements of Acariformes, though never sister to Xiphosura.

```{r trad-bal-comp, out.width="35%", fig.align = "center", warning=FALSE}

#comp = cophylo(trad_tree_19, bal_tree_19)
# Generate the cophylo object

#plot(comp, link.type="curved", link.lwd=4, link.lty="solid", link.col=make.transparent(corecol(numcol=1, offset=3), 0.5), fsize=c(0.5,0.5))
# Plot the trees

#RF.dist(trad_tree_19, bal_tree_19)
comparePhylo(trad_tree_19, bal_tree_19, commons=F, plot=T, location=NA)

```



[< Back to project](index.html)
